onboarding_project/squad_pantry_app/admin.py

```diff
@@ -22,11 +25,87 @@ class OrderDishInline(admin.StackedInline):
     formset = BaseOrderDishFormset
     extra = 1
 
+    def formfield_for_foreignkey(self, db_field, request, **kwargs):
+        if db_field.name == "dish":
+            kwargs["queryset"] = Dish.objects.filter(is_available=True)
+        return super().formfield_for_foreignkey(db_field, request, **kwargs)
+
+    def get_readonly_fields(self, request, obj=None):
+        try:
+            if obj.placed_by:
+                return self.readonly_fields + ('dish', 'quantity')
+        except AttributeError:
+            return self.readonly_fields
+
+
+class DishAdmin(admin.ModelAdmin):
+    list_display = ['dish_name', 'dish_type', 'is_available', 'prep_time_in_minutes']
+
+    def has_add_permission(self, request):
+        return request.user.is_kitchen_staff
+
+    def has_change_permission(self, request, obj=None):
+        return request.user.is_kitchen_staff
+
 
 class OrderAdmin(admin.ModelAdmin):
-    inlines = [OrderDishInline,]
+    inlines = [OrderDishInline, ]
+    exclude = ('placed_by', )
+    readonly_fields = ('closed_at', )
+    list_display = ('placed_by', 'status', 'created_at')
+
+    def get_urls(self):
+        return [
+            url(
+                r'^(.+)/change/cancel-order/$',
+                self.admin_site.admin_view(self.cancel_order_view),
+                name='cancel_order_view',
+            ),
+        ] + super(OrderAdmin, self).get_urls()
+
+    def cancel_order_view(self, request, object_id, extra_context=None):
+
+        try:
+            order = Order.objects.get(pk=object_id)
+        except Order.DoesNotExist:
+            self.message_user(request, 'Mentioned Order ID does not exists', messages.ERROR)
+            return HttpResponseRedirect(request.META.get('HTTP_REFERER', '/admin/squad_pantry_app/order'))
+
+        cancelled = order.cancel_order(request.user)
+        if cancelled == order.CANCEL_SUCCESS:
+            self.message_user(request, 'Order cancelled', messages.SUCCESS)
+        elif cancelled == order.WRONG_USER:
+            self.message_user(request, 'You did not place this order', messages.ERROR)
+        elif cancelled == order.PROCESSING_ORDER:
+            self.message_user(request, 'Could not cancel, The order is already being processed', messages.ERROR)
+        elif cancelled == order.ORDER_CLOSED_ERROR:
+            self.message_user(request, 'The order is already closed', messages.ERROR)
+
+        return HttpResponseRedirect(request.META.get('HTTP_REFERER', '/'))
```

Comment:
In Squad code base we usually do a `HttpResponseRedirect(reverse('admin:squad_pantry_app_order_change_{0}'.format(object_id))`
I do not know what we should use here. Can you spend 10 minutes to understand the difference between the `reverse()` approach and `HTTP_REFERER` approach? Figure out the pros and cons of each.

Reply:
Using reverse is giving this error!
![screen shot 2018-02-06 at 11 58 53 am](https://user-images.githubusercontent.com/13600114/35844770-2d8aae74-0b35-11e8-803b-39dafc510690.png)

Reply:
Try `reverse('admin:squad_pantry_app_order_change', args=(3,), current_app=self.admin_site.name)`. You should still tell me what's the difference between both approaches, if there _is_ a difference at all.